atl-checker/src/algorithms/certain_zero/search_strategy/linear_programming_search.rs

```diff
@@ -0,0 +1,224 @@
+
+            for constraint in &constraints {
+                if constraint.comparison == ComparisonOp::NotEqual {
+                    // We skip not-equal constraints since the minilp cannot handle them.
+                    // They don't restrict the solution space in a meaning way anyway.
+                    continue;
+                }
+                // Build constraint
+                let mut lin_expr = LinearExpr::empty();
+                // The contraint is offset by the given state. This is reflected by adding
+                // coefficient*value_of_symbol_in_state to the constant
+                let mut offset = -constraint.constant;
+                for (symbol, coefficient) in &constraint.terms {
+                    // Get symbol variable or register if missing
+                    let var = symbol_vars.entry(symbol.clone()).or_insert_with(|| {
+                        let decl = self.game.get_decl(symbol).unwrap();
+                        if let DeclKind::StateVar(var_decl) = &decl.kind {
+                            let range = (
+                                *var_decl.ir_range.start() as f64,
+                                *var_decl.ir_range.end() as f64,
+                            );
+                            problem.add_var(1.0, range)
+                        } else {
+                            panic!("Proposition contains a non-variable")
+                        }
+                    });
+                    lin_expr.add(*var, *coefficient);
+                    offset += *coefficient * state.0[&symbol] as f64;
+                }
+
+                problem.add_constraint(lin_expr, constraint.comparison.into(), offset);
+            }
+
+            if let Ok(solution) = problem.solve() {
+                // Extract solution
+                let mut sol_state: HashMap<SymbolIdentifier, i32> = HashMap::new();
+                let mut man_dist = 0;
+                for (symbol, var) in symbol_vars {
+                    let v = solution[var] as i32;
+                    man_dist += (state.0[&symbol] - v).abs();
+                    sol_state.insert(symbol, v);
+                }
+
+                if let Some((_, old_best_dist)) = best {
+                    if old_best_dist > man_dist {
+                        // We found a better solution
+                        best = Some((State(sol_state), man_dist));
+                    }
+                } else {
+                    // We our first possible a solution
+                    best = Some((State(sol_state), man_dist));
+                }
+            }
+        }
+
+        best
+    }
+}
+
+/// Returns all variants of linear problems that can be created from the given LinearConstrainedPhi.
+/// Each problem consists of a list of linear constraints.
+fn all_variants(phi: &LinearConstrainedPhi) -> Vec<Vec<LinearConstraint>> {
+    all_variants_rec(vec![vec![]], phi)
+}
+/// Recursive helper function to [all_variants].
+fn all_variants_rec(
```

Comment:
Also, not quite sure I grasp exactly how this works. But essentially it becomes a vec of all possible combinations of the ways we could satisfy the phi?

Reply:
Yes. So if p, q, r, s, t are constraints then `(p OR q) AND (r OR (s AND t))` becomes `[p+r, q+r, p+s+t, q+s+t]`. It is similar to generating all words in the context free grammar
```
S -> AB
A -> p | q
B -> r | st
```
Some reference method for doing this: https://math.stackexchange.com/questions/36403/generating-all-words-in-a-language-from-cfg